clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,223 @@
+// RUN: %clang_cc1 %s --std=c++17 -fsyntax-only -fsycl-is-device -Wno-unused -verify
+
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) { // #kernelSingleTask
+  kernelFunc();
+}
+
+// kernel3 - expect error
+// The current function is named with a lambda (i.e., takes a lambda as a
+// template parameter. Call the builtin on the current function
+// Then it is passed to kernel
+// Error is thrown because the current function is passed to the kernel,
+// thus part of the latter's name, after its unique stable is already computed
+// Current function is a part of that name, causing its unique_stable_name
+// mangling to change
+template <typename Func>
+void kernel3func(const Func &F3) {
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{'__builtin_unique_stable_name' evaluated here}}
+  constexpr const char *F3_output = __builtin_unique_stable_name(F3);
+
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#kernel3func_call{{in instantiation of function template specialization}}
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel3>(F3);
+}
+
+// kernel4 - expect error
+// The current function is named with a lambda (i.e., takes a lambda as a
+// template parameter. Call the builtin on the current function
+// Then an empty lambda is passed to kernel
+// Error is thrown because the empty lambda is named based on kernel4func
+// Current function is a part of that name, causing its unique_stable_name
+// mangling to change
+template <typename Func>
+void kernel4func(const Func &F4) {
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_F4{{'__builtin_unique_stable_name' evaluated here}}
+  constexpr const char *F4_output = __builtin_unique_stable_name(F4); // #USN_F4
+
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#kernel4func_call{{in instantiation of function template specialization}}
+  // expected-note@#USN_l7{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel4>([]() {});
+}
+
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct T {};
```

Comment:
Oh please, never name a type 'T'.  That is perhaps the most confusing thing that has happened to me in a long while...

Reply:
Yikes! Sorry.

Reply:
Can you rename `struct T` to something else, like `struct Tangerine`?